File: CS/WindowsFormsApplication122/Form1.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Windows.Forms;
using DevExpress.XtraReports.Design;
using DevExpress.XtraReports.Design.Commands;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.UserDesigner;
// ...

namespace WindowsFormsApplication122 {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }
        private void button1_Click(object sender, EventArgs e) {
            XtraReport report = new XtraReport();
            XRDesignForm form = new XRDesignForm();
            form.DesignMdiController.AddService(typeof(IMenuCreationService),
                new CustomMenuCreationService(form.DesignMdiController));
            form.OpenReport(report);
            form.ShowDialog();
        }
    }
    class CustomMenuCreationService : IMenuCreationService {
        XRDesignMdiController controller;
        XtraReport ActiveReport {
            get {
                return controller.ActiveDesignPanel != null ? controller.ActiveDesignPanel.Report : null;
            }
        }
        CommandID showGridCommandID;
        CommandID hideGridCommandID;
        public CustomMenuCreationService(XRDesignMdiController controller) {
            this.controller = controller;
            Guid guid = Guid.NewGuid();
            showGridCommandID = new CommandID(guid, 0);
            hideGridCommandID = new CommandID(guid, 1);
        }
        public void ProcessMenuItems(MenuKind menuKind, MenuItemDescriptionCollection items) {
            if (menuKind == MenuKind.Selection) {
                int index = items.IndexOf(ReportCommand.Copy) - 1;
                if (index < 0)
                    return;
                MenuItemDescription itemShow = new MenuItemDescription("Show Grid", null, showGridCommandID);
                MenuItemDescription itemHide = new MenuItemDescription("Hide Grid", null, hideGridCommandID);
                items.Insert(index++, MenuItemDescription.Separator);
                items.Insert(index++, itemShow);
                items.Insert(index++, itemHide);
                items.Insert(index++, MenuItemDescription.Separator);
            }
        }
        public MenuCommandDescription[] GetCustomMenuCommands() {
            return new MenuCommandDescription[] { new MenuCommandDescription(showGridCommandID, OnHandleShowGrid, OnStatusShowGrid),
                new MenuCommandDescription(hideGridCommandID, OnHandleHideGrid, OnStatusHideGrid)};
        }
        void OnHandleShowGrid(object sender, CommandExecuteEventArgs e) {
            SetDrawGrid(true);
        }
        void OnHandleHideGrid(object sender, CommandExecuteEventArgs e) {
            SetDrawGrid(false);
        }
        void SetDrawGrid(bool value) {
            PropertyDescriptor property = TypeDescriptor.GetProperties(typeof(XtraReport))["DrawGrid"];
            property.SetValue(controller.ActiveDesignPanel.Report, value);
        }
        void OnStatusShowGrid(object sender, EventArgs e) {
            MenuCommand command = sender as MenuCommand;
            command.Supported = command.Enabled = ActiveReport != null && !ActiveReport.DrawGrid;
        }
        void OnStatusHideGrid(object sender, EventArgs e) {
            MenuCommand command = sender as MenuCommand;
            command.Supported = command.Enabled = ActiveReport != null && ActiveReport.DrawGrid;
        }
    }
}

File: VB/E4343/Form1.vb
Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Windows.Forms
Imports DevExpress.XtraReports.Design
Imports DevExpress.XtraReports.Design.Commands
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.UserDesigner
' ...

Namespace E4343
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
		End Sub
		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			Dim report As New XtraReport()
			Dim form As New XRDesignForm()
			form.DesignMdiController.AddService(GetType(IMenuCreationService), New CustomMenuCreationService(form.DesignMdiController))
			form.OpenReport(report)
			form.ShowDialog()
		End Sub
	End Class
	Friend Class CustomMenuCreationService
		Implements IMenuCreationService

		Private controller As XRDesignMdiController
		Private ReadOnly Property ActiveReport() As XtraReport
			Get
				Return If(controller.ActiveDesignPanel IsNot Nothing, controller.ActiveDesignPanel.Report, Nothing)
			End Get
		End Property
		Private showGridCommandID As CommandID
		Private hideGridCommandID As CommandID
		Public Sub New(ByVal controller As XRDesignMdiController)
			Me.controller = controller
			Dim guid As Guid = System.Guid.NewGuid()
			showGridCommandID = New CommandID(guid, 0)
			hideGridCommandID = New CommandID(guid, 1)
		End Sub
		Public Sub ProcessMenuItems(ByVal menuKind As MenuKind, ByVal items As MenuItemDescriptionCollection) Implements IMenuCreationService.ProcessMenuItems
			If menuKind = DevExpress.XtraReports.UserDesigner.MenuKind.Selection Then
				Dim index As Integer = Math.Max(0, items.IndexOf(ReportCommand.Cut))
				Dim itemShow As New MenuItemDescription("Show Grid", Nothing, showGridCommandID)
				Dim itemHide As New MenuItemDescription("Hide Grid", Nothing, hideGridCommandID)
				items.Insert(index, MenuItemDescription.Separator)
				index += 1
				items.Insert(index, itemShow)
				index += 1
				items.Insert(index, MenuItemDescription.Separator)
				index += 1
				items.Insert(index, itemHide)
				index += 1
				items.Insert(index, MenuItemDescription.Separator)
				index += 1
			End If
		End Sub
		Public Function GetCustomMenuCommands() As MenuCommandDescription() Implements IMenuCreationService.GetCustomMenuCommands
            Return New MenuCommandDescription() {
                New MenuCommandDescription(showGridCommandID, AddressOf OnHandleShowGrid, AddressOf OnStatusShowGrid),
                New MenuCommandDescription(hideGridCommandID, AddressOf OnHandleHideGrid, AddressOf OnStatusHideGrid)
            }
        End Function
        Private Sub OnHandleShowGrid(ByVal sender As Object, ByVal e As CommandExecuteEventArgs)
            SetDrawGrid(True)
        End Sub
        Private Sub OnHandleHideGrid(ByVal sender As Object, ByVal e As CommandExecuteEventArgs)
            SetDrawGrid(False)
        End Sub
        Private Sub SetDrawGrid(ByVal value As Boolean)
            Dim [property] As PropertyDescriptor = TypeDescriptor.GetProperties(GetType(XtraReport))("DrawGrid")
            [property].SetValue(controller.ActiveDesignPanel.Report, value)
        End Sub
        Private Sub OnStatusShowGrid(ByVal sender As Object, ByVal e As EventArgs)
            Dim command As MenuCommand = TryCast(sender, MenuCommand)
            command.Enabled = ActiveReport IsNot Nothing AndAlso Not ActiveReport.DrawGrid
            command.Supported = command.Enabled
        End Sub
        Private Sub OnStatusHideGrid(ByVal sender As Object, ByVal e As EventArgs)
            Dim command As MenuCommand = TryCast(sender, MenuCommand)
            command.Enabled = ActiveReport IsNot Nothing AndAlso ActiveReport.DrawGrid
            command.Supported = command.Enabled
        End Sub
    End Class
End Namespace
